### PR TITLE
perf(circom): optimize zkey wtns reading

### PR DESCRIPTION
**Pull request by chokobole**
# Description

Zkey and wtns files are now read using `mmap()` and is much faster!

Before

```
Start parsing zkey
Time taken for parsing zkey: 5.97333 s
Start parsing witness
Time taken for parsing witness: 0.319247 s
```

After
```
Start parsing zkey
Time taken for parsing zkey: 0.611177 s
Start parsing witness
Time taken for parsing witness: 0.097334 s
```

### CODE REVIEW

**Review by dongchangYoo — APPROVED** [review]
LGTM, please check linter error!

**Review by GideokKim — APPROVED** [review]
LGTM

File: tachyon/base/files/memory_mapped_file.cc [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "tachyon/base/files/memory_mapped_file.h"
+
+#include <utility>
+
+#include "tachyon/base/files/file_path.h"
+#include "tachyon/base/logging.h"
+// #include "tachyon/base/notreached.h"
```
Comment (ashjeong) [inline tachyon/base/files/memory_mapped_file.cc:11]: Can this be removed?

File: tachyon/base/files/memory_mapped_file.cc [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  return !bytes_.empty();
+}
+
+// static
+void MemoryMappedFile::CalculateVMAlignedBoundaries(int64_t start,
+                                                    size_t size,
+                                                    int64_t* aligned_start,
+                                                    size_t* aligned_size,
+                                                    int32_t* offset) {
+  // Sadly, on Windows, the mmap alignment is not just equal to the page size.
```
Comment (ashjeong) [inline tachyon/base/files/memory_mapped_file.cc:135]: Not sure if this conveys what you want
```suggestion
  // On Windows, the mmap alignment is not necessarily equal to the page size.
```

**Review by batzor — APPROVED** [review]
LGTM

File: tachyon/base/files/memory_mapped_file.cc [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  uint64_t mask = SysInfo::VMAllocationGranularity() - 1;
+  CHECK(IsValueInRangeForNumericType<int32_t>(mask));
+  *offset = static_cast<int32_t>(static_cast<uint64_t>(start) & mask);
+  *aligned_start = static_cast<int64_t>(static_cast<uint64_t>(start) & ~mask);
+  // The CHECK above means bit 31 is not set in `mask`, which in turn means
+  // *offset is positive.  Therefore casting it to a size_t is safe.
```
Comment (ashjeong) [inline tachyon/base/files/memory_mapped_file.cc:141]:
```suggestion
  // The CHECK above means bit 31 is not set in `mask`, which in turn means
  // *offset is positive. Therefore, casting it to a size_t is safe.
```

File: tachyon/base/files/memory_mapped_file.h [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+#include <stddef.h>
+#include <stdint.h>
+
+#include <utility>
+
+#include "absl/types/span.h"
+
+#include "tachyon/export.h"
+#include "tachyon/base/files/file.h"
+// #include "tachyon/base/memory/raw_ptr_exclusion.h"
```
Comment (ashjeong) [inline tachyon/base/files/memory_mapped_file.h:17]: Can we remove this line?

File: tachyon/base/files/memory_mapped_file_posix.cc [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+#include <stddef.h>
+#include <stdint.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "tachyon/base/files/file_util.h"
+#include "tachyon/base/logging.h"
+#include "tachyon/base/numerics/safe_conversions.h"
+// #include "tachyon/base/threading/scoped_blocking_call.h"
```
Comment (ashjeong) [inline tachyon/base/files/memory_mapped_file_posix.cc:17]: Can we remove this line?

File: tachyon/base/files/memory_mapped_file_posix.cc [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+      return false;
+    }
+
+    map_start = base::checked_cast<off_t>(aligned_start);
+    map_size = aligned_size;
+    byte_size = region.size;
+  }
+
+  if (map_size == 0u) {
+    // mmap() requires `map_size > 0`, and this ensures an empty span indicates
```
Comment (ashjeong) [inline tachyon/base/files/memory_mapped_file_posix.cc:72]: Is this taken from Chromium (aka we should leave it alone? Else, please wrap names with "|")

**Review by ashjeong — COMMENTED** [review]
https://github.com/kroma-network/tachyon/commit/78af3cc9916af152413bd07852afefb955d24f60
Please change "belows" to "below" in the commit message!

File: tachyon/base/files/memory_mapped_file_posix.cc (hunk 6 above)
Comment (chokobole) [reply tachyon/base/files/memory_mapped_file_posix.cc:72]: Yes it was taken from chromium! is it okay to keep it as it is for maintenance?

File: tachyon/base/files/memory_mapped_file_posix.cc (hunk 5 above)
Comment (chokobole) [reply tachyon/base/files/memory_mapped_file_posix.cc:17]: I'll remain this line to minimize diff and remind me later of uncommenting after we take `ScopedBlockingCall`.

File: tachyon/base/files/memory_mapped_file.h (hunk 4 above)
Comment (chokobole) [reply tachyon/base/files/memory_mapped_file.h:17]: I'll remain this line to minimize diff and remind me later of uncommenting after we take `raw_ptr_exclusion.h`.

File: tachyon/base/files/memory_mapped_file.cc (hunk 3 above)
Comment (chokobole) [reply tachyon/base/files/memory_mapped_file.cc:141]: This is taken from chromium and i want it to be untouched for maintenance issue

**Review by ashjeong — APPROVED** [review]
LGTM please fix ci lint~

**Comment by chokobole** [comment]
> LGTM please fix ci lint~

CI lint asks me to fix the code from chromium, so I'll leave it as it is for easy maintenance.

**Review by TomTaehoonKim — APPROVED** [review]
Else, LGTM

**Review by batzor — APPROVED** [review]
LGTM

File: tachyon/base/files/file_util_posix.cc [hunk 7]
```diff
@@ -1408,7 +1408,7 @@ TACHYON_EXPORT bool IsPathExecutable(const FilePath& path) {
     CHECK_GE(sysconf_result, 0);
     size_t pagesize = static_cast<size_t>(sysconf_result);
     CHECK_GE(sizeof(pagesize), sizeof(sysconf_result));
-    void* mapping = mmap(nullptr, pagesize, PROT_READ, MAP_SHARED, fd.get(), 0);
+    void* mapping =    mmap(nullptr, pagesize, PROT_READ, MAP_SHARED, fd.get(), 0);
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/base/files/file_util_posix.cc]: Oh it was my bad, thanks for finding!

**Review by GideokKim — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
It looks much cleaner now. LGTM again.

**Pull request merged**